design/ceph/object/swift-and-keystone-integration.md

```diff
@@ -0,0 +1,241 @@
+  to the bucket and use those credentials.
+
+* Support for authentication technologies other than Keystone (e.g. LDAP)
+
+* Exposing options that disable security features (e.g. TLS verification)
+
+## Proposal details
+
+The Object Store CRD will have to be extended to accommodate the new
+settings.
+
+### Keystone integration
+
+A new section `auth:` is added to the Object Store CRD. To configure
+the keystone integration:
+
+```yaml
+auth:
+  keystone:
+    url: https://keystone:5000/
+    acceptedRoles: ["_member_", "service", "admin"]
+    apiVersion: 3
+    implicitTenants: swift
+    tokenCacheSize: 1000
+    revocationInterval: 1200
+    serviceUserSecret: rgw-service-user
+```
+
+The certificate to verify the Keystone endpoint can't be explicitly
+configured in Ceph RGW, the system configuration of the pod is used,
+you can add to the system certificate store via the
+`gateway.caBundleRef` setting of the object store CRD.
+
+The credentials for the Keystone service account used by Ceph RGW are
+supplied in a Secret that contains a mapping of OpenStack openrc
+environment variables. Only password authentication to Keystone is
+supported. Example:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: rgw-service-user
+data:
+  OS_PASSWORD: "horse staples battery correct"
+  OS_USERNAME: "ceph-rgw"
+  OS_PROJECT_NAME: "admin"
+  OS_USER_DOMAIN_NAME: "Default"
+  OS_PROJECT_DOMAIN_NAME: "Default"
+  OS_AUTH_TYPE: "password"
+```
+
+This format is chosen because it is a natural and interoperable way
+that keystone credentials are represented and for the supported auth
+type it maps naturally to the Ceph RGW configuration.
+
+The following constraints must be fulfilled by the secret:
+* `OS_AUTH_TYPE` must be `password` or omitted.
+* `OS_USER_DOMAIN_NAME` must equal `OS_PROJECT_DOMAIN_NAME`.
+* All openrc variables not in the example above (e.g. API version and
+  Keystone endpoint) are ignored.
+
+### Swift integration
+
+Swift is configured in the `gateway:` section of the Object Store CRD.
+
+The currently unused `type:` argument may now take the values `swift`
+or `s3`.
+
+The following new settings are available to properly configure Swift:
+```yaml
+gateway:
+  type: swift
+  swiftAccountInUrl: true
+  swiftUrlPrefix: /swifter
+  swiftVersioningEnabled: false
+```
+
+The access to the Swift API is granted by creating a subuser of an RGW
+user. While commonly the access is granted via projects
+mapped from Keystone, explicit creation of subusers is supported by
+extending the `cephobjectstoreuser` resource with a new section
+`spec.subUsers`:
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephObjectStoreUser
+metadata:
+  name: my-user
+  namespace: rook-ceph
+spec:
+  store: my-store
+  displayName: my-display-name
+  quotas:
+    maxBuckets: 100
+    maxSize: 10G
+    maxObjects: 10000
+  capabilities:
+    user: "*"
+    bucket: "*"
+  subUsers:
+  - name: swift
+    access: full
```

Comment:
What are the possible values for `access`? Generally, design docs comment on newly added fields with `[N]` comments and then have a numbered list below to discuss the fields in detail.

Reply:
Fixed.